lxd/instance/drivers/driver_qemu.go

```diff
@@ -2929,19 +2942,42 @@ func (d *qemu) IsPrivileged() bool {
 
 // Snapshot takes a new snapshot.
 func (d *qemu) Snapshot(name string, expiry time.Time, stateful bool) error {
+	// Deal with state.
 	if stateful {
-		return fmt.Errorf("Can't perform a stateful snapshot of a virtual machine")
+		// Confirm the instance has stateful migration enabled.
+		if !shared.IsTrue(d.expandedConfig["migration.stateful"]) {
+			return fmt.Errorf("Stateful stop requires migration.stateful to be set to true")
+		}
+
+		// Sanity checks.
+		if !d.IsRunning() {
+			return fmt.Errorf("Unable to create a stateful snapshot. The instance isn't running")
+		}
+
+		// Connect to the monitor.
+		monitor, err := qmp.Connect(d.monitorPath(), qemuSerialChardevName, d.getMonitorEventHandler())
+		if err != nil {
+			return err
+		}
+
+		// Dump the state.
+		err = d.saveState(monitor)
+		if err != nil {
+			return err
+		}
+
+		// Resume the VM once the disk state has been saved.
+		defer monitor.Start()
```

Comment:
@stgraber why does the monitor need restarting here, does something in d.saveState() stop it? If so should that be handled in d.saveState() instead of here?

Reply:
@stgraber ah, is it that d.saveState pauses the VM and we dont want to resume it until after the snapshot is taken?

Reply:
@stgraber what actually causes the VM to pause here (I looked in saveState but cannot see it).

Reply:
`monitor.Start()` sends `cont` as a command over QMP which resumes the virtualization.
Otherwise at the end of the state dump, the virtualization is paused and VM is unresponsive.

It's handled here because with stateful snapshots we need the VM to run again after the snapshot is taken.
When we do a stateful stop or migration, we specifically don't want it to do that and instead use `monitor.Quit` to completely kill it.

Reply:
@stgraber got it thanks. Maybe the function could be called `saveStateAndPause` instead, or just update the comment to explain that running the function will result in the VM being paused?

Reply:
I guess we can add a comment.

To me it felt obvious that any such function would result in the workload being stopped as otherwise you would immediately get out of sync with the disk state, not to mention confuse the hell out of network connections.

Reply:
@stgraber yes its just my lack of familiarity with it, I thought it might be like our snapshot functionality, where we take the snapshot and then resume the instance, and not leave it stopped/paused.

Reply:
updated